### PR TITLE
[labs/analyzer] Adds support for analyzing JavaScript packages

### PR DESCRIPTION
**Pull request by kevinpschaaf**
* Removed the automatic tsconfig file lookup in the `packageRoot` to instead require `packageRoot` to _either_ point to a tsconfig file, or else a package root containing a `tsconfig.json`; if none is found, we assume a JS program. This avoids `ts.findConfigFile` from searching to the root of your hard disk when analyzing a JS program; seems better to be explicit anyway.
* If no tsconfig is found, we create a barebones TS program using `allowJS` and a glob of all `*.js` within the `packageRoot`
* Added support for JS-specific analysis:
  * `customElements.define()`
  * `static properties = {...}` and `static get properties() { return {...} }`
  * Inference of types initialized in the constructor (note even in a JS program, TS does a decent job at inference)
  * Types on the `ReactiveProperty` model is now optional in case there is no initializer we can even infer `any` from
  * Add `isOptional` and `isNonNull` to the `ReactiveProperty` model (based on `?` and `!` post-fix operators on the declarations, or a best-effort JS fallback); this info was being used in the Vue generator
* Duplicate / de-TypeScript-ify all the `test-files` (input source for tests) under separate `js` / `ts` folders
* Run all the unit tests in `for (const lang of ['js', 'ts'])` loop

### CODE REVIEW

File: packages/labs/analyzer/src/test/analyzer_test.ts [hunk 1]
```diff
@@ -12,41 +12,52 @@ import {fileURLToPath} from 'url';
...
+  const test = suite<{analyzer: Analyzer; packagePath: AbsolutePath}>(
+    `Basic Analyzer tests (${lang})`
   );
-  assert.equal(elementAModule?.jsPath, path.normalize('out/class-a.js'));
-  assert.equal(elementAModule?.declarations.length, 1);
-  assert.equal(elementAModule?.declarations[0].name, 'ClassA');
-});
 
-test.run();
+  const getSourceFilename = (f: string) =>
```
Comment (sorvell) [inline packages/labs/analyzer/src/test/analyzer_test.ts:20]: Pull these out into test utilities?

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 2]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+ */
+const addPropertiesFromStaticBlock = (
+  classDeclaration: LitClassDeclaration,
+  properties: ts.PropertyDeclaration | ts.GetAccessorDeclaration,
+  undecoratedProperties: Map<string, ts.Node>,
+  reactiveProperties: Map<string, ReactiveProperty>,
+  analyzer: AnalyzerInterface
+) => {
+  // Add any constructor initializers to the undecorated properties node map
+  // from which we can infer types from. This is the primary path that JS source
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts:99]: I can try and clarify in the comment. Assuming native "use define for class field semantics", you can't initialize the property using a native field initializer, you need to do it in the constructor. However, You _can't_ do that in TypeScript without also having a `declare foo: string` or whatever, so this extra "look in the constructor" path is _primarily_ for JS initializers of reactive properties.

If a JS user disabled the "use define for class fields" in their compiler, then they will already have entries in the `undecoratedProperties` map (and we don't overwrite that with constructor initializers).

File: packages/labs/analyzer/src/lib/analyze-package.ts [hunk 3]
```diff
@@ -6,28 +6,46 @@ import {Analyzer} from './analyzer.js';
...
-  const configFileName = ts.findConfigFile(
-    packageRoot,
-    ts.sys.fileExists,
-    'tsconfig.json'
-  );
-  if (configFileName === undefined) {
-    // TODO: use a hard-coded tsconfig for JS projects.
-    throw new Error(`tsconfig.json not found in ${packageRoot}`);
+export const createPackageAnalyzer = (packagePath: AbsolutePath) => {
+  // This logic accepts either a path to folder containing a tsconfig.json
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/analyze-package.ts:15]: Done

File: packages/labs/analyzer/src/lib/analyze-package.ts [hunk 4]
```diff
@@ -6,28 +6,46 @@ import {Analyzer} from './analyzer.js';
...
+      packagePath /* basePath */,
+      undefined /* existingOptions */,
+      path.relative(packagePath, configFileName) /* configFileName */
+    );
+  } else if (isDirectory) {
+    console.info(`No tsconfig.json found; assuming package is JavaScript.`);
+    commandLine = ts.parseJsonConfigFileContent(
+      {
+        compilerOptions: {
+          module: 'esm',
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/analyze-package.ts:38]: No, will add a TODO; should probably be configurable.

File: packages/labs/analyzer/src/lib/analyzer.ts [hunk 5]
```diff
@@ -77,17 +84,30 @@ export class Analyzer implements AnalyzerInterface {
...
+    return ts.parseJsonConfigFileContent(
+      json,
+      ts.sys,
+      analyzer.path.basename(compilerOptions.configFilePath as string),
+      undefined,
+      compilerOptions.configFilePath as string
+    );
+  } else {
+    // Otherwise, this is a JS project; we need to determine the package root
+    const packageRoot = getPackageRootForModulePath(
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/analyzer.ts:112]: DOne

File: packages/labs/analyzer/src/lib/lit-element/lit-element.ts [hunk 6]
```diff
@@ -107,7 +106,32 @@ export const getTagName = (declaration: LitClassDeclaration) => {
     customElementDecorator.expression.arguments.length === 1 &&
     ts.isStringLiteral(customElementDecorator.expression.arguments[0])
   ) {
-    tagname = customElementDecorator.expression.arguments[0].text;
+    // Get tag from decorator
+    tagName = customElementDecorator.expression.arguments[0].text;
+  } else {
+    // Otherwise, search for customElements.define
+    declaration.parent.forEachChild((child) => {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/lit-element.ts:114]: Done

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 7]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+      // Note we will process this after the loop so that the
+      // `undecoratedProperties` map is complete before processing the static
+      // properties block.
+      staticProperties = prop;
+    } else {
+      // Store the declaration node for any undecorated properties. In a TS
+      // program that happens to use a static properties block along with
+      // the `declare` keyword to type the field, we can use this node to
+      // get/infer the TS type of the field from
+      undecoratedProperties.set(name, prop);
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts:67]: Yeah, nice catch. Done.

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 8]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+  analyzer: AnalyzerInterface
+) => {
+  // Add any constructor initializers to the undecorated properties node map
+  // from which we can infer types from. This is the primary path that JS source
+  // can get their inferred types (in TS, types will come from the undecorated
+  // fields passed in, since you need to declare the field to assign it in the
+  // constructor).
+  addConstructorInitializers(classDeclaration, undecoratedProperties);
+  // Find the object literal from the initializer or getter return value
+  const object = getStaticPropertiesObjectLiteral(properties);
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts:103]: Yes

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 9]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+  // Loop over each key/value in the object and add them to the map
+  for (const prop of object.properties) {
+    if (
+      ts.isPropertyAssignment(prop) &&
+      ts.isIdentifier(prop.name) &&
+      ts.isObjectLiteralExpression(prop.initializer)
+    ) {
+      const name = prop.name.text;
+      const options = prop.initializer;
+      const nodeForType = undecoratedProperties.get(name);
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts:113]: Fixed

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 10]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+      // In JS (where we won't expect a PropertyDeclaration), we treat nodes
+      // with a ctor initializer as non-null and optional; in TS when we have
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts]: Removed, as discussed

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 11]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+      // a PropertyDeclaration, we base it on the post-fix operators
+      const isNonNull =
+        nodeForType === undefined
+          ? false
+          : ts.isPropertyDeclaration(nodeForType)
+          ? !!nodeForType.exclamationToken
+          : true;
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts]: Removed, as discussed

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 12]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+      const isOptional =
+        nodeForType === undefined
+          ? false
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts]: Removed, as discussed

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 13]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+  if (
+    ts.isPropertyDeclaration(properties) &&
+    properties.initializer !== undefined &&
+    ts.isObjectLiteralExpression(properties.initializer)
+  ) {
+    // `properties` has a static initializer; get the object from there
+    object = properties.initializer;
+  } else if (ts.isGetAccessorDeclaration(properties)) {
+    // Object was in a static getter: find the object in the return value
+    const statement = properties.body?.statements[0];
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts]: Done.

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 14]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+const addConstructorInitializers = (
+  classDeclaration: ts.ClassDeclaration,
+  undecoratedProperties: Map<string, ts.Node>
+) => {
+  const ctor = classDeclaration.forEachChild((node) =>
+    ts.isConstructorDeclaration(node) ? node : undefined
+  );
+  if (ctor !== undefined) {
+    ctor.body?.statements.forEach((stmt) => {
+      if (
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts:194]: Done

File: packages/labs/analyzer/src/lib/model.ts [hunk 15]
```diff
@@ -210,6 +209,22 @@ export interface ReactiveProperty {
...
+   * by the existence of the `?` post-fix operator on the property declaration.
+   * In JavaScript, properties that are initialized in the constructor
+   * are considered optional.
+   */
+  isOptional: boolean;
+
+  /**
+   * Whether the property is non-null. In TypeScript, this is determined
+   * by the existence of the `!` post-fix operator on the property declaration.
+   * In JavaScript, properties that are initialized in the constructor
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/model.ts]: Removed, as discussed

File: packages/labs/analyzer/src/lib/lit-element/properties.ts [hunk 16]
```diff
@@ -15,40 +15,217 @@ import {LitClassDeclaration} from './lit-element.js';
...
+      const options = prop.initializer;
+      const nodeForType = undecoratedProperties.get(name);
+      // In JS (where we won't expect a PropertyDeclaration), we treat nodes
+      // with a ctor initializer as non-null and optional; in TS when we have
+      // a PropertyDeclaration, we base it on the post-fix operators
+      const isNonNull =
+        nodeForType === undefined
+          ? false
+          : ts.isPropertyDeclaration(nodeForType)
+          ? !!nodeForType.exclamationToken
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/lit-element/properties.ts]: Removed, as discussed

File: packages/labs/analyzer/src/test/analyzer_test.ts (hunk 1 above)
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/test/analyzer_test.ts]: Done

File: packages/labs/analyzer/test-files/js/basic-elements/element-a.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+  static properties = {
+    a: {},
+    b: {reflect: true, type: Number, attribute: 'bbb'},
+  };
+
+  constructor() {
+    super();
+    this.a = '';
+    this.b = 1;
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/test-files/js/basic-elements/element-a.js:24]: Done

**Review by sorvell — APPROVED** [review]
LGTM

**Pull request merged**